### ISSUE TITLE
Allow trailing commas in parameter and argument lists.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#26644]
To improve the maintainability of long argument and parameter lists, we allow the final argument or parameter to be followed by an optional comma. This allows a style where each argument or parameter is written on a line by itself, and moving one item up or down or between functions doesn't require special handling of the final item.
This matches the existing behavior for list and map literals and enum class declarations.

To be precise:
- The grammar is changed to allow a single comma right after a final parameter or argument.
- It does not allow multiple commas between items, a leading comma, or a comma in an empty parameter or argument list.
- The comma goes before the final ')' in parameter lists with no optional parameters, and before the final '}' or ']' in parameter lists with optional parameters.
- The comma goes before the final ')' in argument lists.
- The comma is entirely optional and has no semantic meaning, it's purely a syntactic convenience for a particular style of writing parameter or argument lists. 

Tracking bugs for individual tasks:
- [x] VM (#26645)
- [x] dart2js (#26646)
- [x] Analyzer (#26647)
- [x] Specification (#26648)
- [ ] Style guide (#26649)
- [x] Formatter (dart-lang/dart_style#509)

**Comment by hterkelsen** [comment]
Can you have a trailing comma after the '}' in a parameter list? eg:
```dart
void foo(
    String x,
    List y,
    {bool z: true},
    ) {
}
```

**Comment by lrhn** [comment]
@hterkelsen 
Trailing comma after `}`:
That was not the plan. This is intended for a style where you have one parameter on each line, and multiple parameters, and you can move the individual parameters around among themselves.

It handles the cases where you move the last parameter somewhere else (leaving a dangling comma on the previous line and, maybe, missing a comma after the moved parameter in the new position), or where you add a new last parameter after the existing one (now missing a comma on the previous line).

In your example, you put the single optional parameter on a line including the braces. You can't move that *line* anywhere where it isn't last in a parameter list, so there is no advantage to allowing a comma after it.

Basically it's for anywhere where you *could* write `, anotherparameter` but didn't. After the optional parameter brace isn't such a place.

**Comment by sethladd** [comment]
For the VM and dart2js, do we need to use a flag to opt-in to the feature?

**Comment by hterkelsen** [comment]
No, it should be enabled by default.

**Comment by mit-mit** [comment]
From discussion in #26649 the style guide update will follow later, and from discussion in release meeting this feature is indeed complete for 1.19, so closing.

**Comment by mit-mit** [comment]
Updating changelog in https://codereview.chromium.org/2220963002/
